client/cache/cache_pattern.go

```diff
@@ -0,0 +1,110 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+)
+
+// ValueReader reads data from data source, such as database
+type ValueReader interface {
+	Query(key string) (interface{}, error)
+}
+
+// ValueWriter writes data to data source, such as database
+type ValueWriter interface {
+	Update(key string, val interface{}) error
+}
+
+// CacheAside is a decorator of Cache, it implements "cache aside" pattern
+type CacheAside struct {
+	Cache
+	reader   ValueReader
+	writer   ValueWriter
+	lifeSpan time.Duration
+}
+
+func NewCacheAside(cache Cache, reader ValueReader, writer ValueWriter, lifeSpan time.Duration) *CacheAside {
+	return &CacheAside{Cache: cache, reader: reader, writer: writer, lifeSpan: lifeSpan}
+}
+
+func (c *CacheAside) Get(ctx context.Context, key string) (interface{}, error) {
+	if val, _ := c.Cache.Get(ctx, key); val != nil {
+		return val, nil
+	}
+
+	val, err := c.reader.Query(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return val, c.Cache.Put(ctx, key, val, c.lifeSpan)
+}
+
+func (c *CacheAside) GetMulti(ctx context.Context, keys []string) ([]interface{}, error) {
```

Comment:
I think add a new method named "QueryBatch" into `ValueReader` would be better. Users always expect that the `CacheAside` work the same way as `Cache`.

For example, users create the `CacheAside` instance by passing a redis Cache instance which read all values  from redis server in one command `MGET`.

But this method change the behavior.